### ISSUE TITLE
Fields to store in DynamoDB

### ISSUE THREAD

**Issue by adamyy** [castorini/anlessini#7]
Currently, the experiment has been completed for `AclAnthology` collection. And we are storing more than just docid and contents in DynamoDB:

![image](https://user-images.githubusercontent.com/14900215/93709208-fe524400-fb6e-11ea-91b7-262464a4f005.png)

Are these fields arbitrarily decided? Or are they tied to the front end logic? @shaneding It looks like these fields are directly returned to the frontend in the form of JSON strings. https://github.com/castorini/anlessini/blob/master/SearchLambdaFunction/src/main/java/io/anlessini/SearchLambda.java#L112

I am not very comfortable with bundling frontend logic with backend storage structure directly. For instance, what happens if we decided to add another field in the UI? That would have involved a database migration in the current implementation.

Alternatively, what if we just store the raw contents and the ids in Dynamo, and query for additional fields from the index? The front end can specify the fields it want, and the lambda can answer them by doing `searcher.doc(docid).getFields(fieldnames)`. The performance implication of such strategy is yet to be examined.

**Comment by lintool** [comment]
We have nice collection abstractions in Anserini: https://github.com/castorini/anserini/tree/master/src/main/java/io/anserini/collection

A clean solution would be just to use those abstractions. Downside is that they're in Java... although we've exposed in Pyserini: https://github.com/castorini/pyserini/blob/master/docs/usage-collection.md (but not well tested)...

On the search end, we assume this as default: https://github.com/castorini/anserini/blob/master/src/main/java/io/anserini/index/generator/DefaultLuceneDocumentGenerator.java

That the Lucene scheme has at least these three args: https://github.com/castorini/anserini/blob/master/src/main/java/io/anserini/index/IndexArgs.java#L24

So tl;dr - I don't mind if Anlessini is serverless _Anserini_ (hence the name), as opposed to general serverless Lucene...

**Comment by shaneding** [comment]
Sorry I am a bit confused @lintool are you proposing to ditch DynamoDB entirely, and just fetch content via collections? Also @adamyy I agree with what you said, we can have just an `ID` field and then `RAW` in DynamoDB. But I don't think querying directly from the index is viable, since we wanted to use DynamoDB in the first place so that we didn't have to store raw text in the indexes, because that would dramatically increase the size of the indexes that we are fetching from S3.

**Comment by lintool** [comment]
Nope, documents should go into DynamoDB - my proposal is to use Anserini abstractions to insert into DynamoDB and retrieve from it, so user never needs deal with DynamoDB abstractions directly.

**Comment by adamyy** [comment]
So given a collection, we should use Anserini abstractions such as `Collection` and `Generator` to generate the Lucene `Document` and extract the fields information, before push those fields in DynamoDB. Both Lucene and DynamoDB are sort of "schema-less" and we should not have much trouble juggling between the Lucene `Document` and the Dynamo `Item`.

@lintool Is that correct?

**Comment by adamyy** [comment]
@shaneding I see your point. So for Anlessini, the point is to have as little **stored** fields in the index as possible (possibly none), and only use the index to retrieve docids and ranks. Then the docids are used to query from Dynamo to get the actual fields. Correct?

**Comment by shaneding** [comment]
@adamyy I think @lintool meant that we combine the insertion into dynamoDB into the document generation process in Anserini instead of doing it in Python? Correct me if I am wrong. And yeah, our goal is to store as little as possible in S3, and then fetch the bulk of the content from DynamoDB, after searching the index stored in S3.

**Comment by lintool** [comment]
Correct. Doesn't need to be Anserini "proper". Could be a Python script in this repo that `from pyserini import ...`.
